runtime/lua/vim/lsp/client.lua

```diff
@@ -0,0 +1,663 @@
+--- will always be `false` (the client has shutdown).
+--- If {status} is `true`, the function returns {request_id} as the second
+--- result. You can use this with `client.cancel_request(request_id)` to cancel
+--- the request.
+--- @field request fun(method: string, params: table?, handler: lsp.Handler?, bufnr: integer): boolean, integer?
+---
+--- Sends a request to the server and synchronously waits for the response.
+--- This is a wrapper around {client.request}
+--- Returns: { err=err, result=result }, a dictionary, where `err` and `result`
+--- come from the |lsp-handler|. On timeout, cancel or error, returns `(nil,
+--- err)` where `err` is a string describing the failure reason. If the request
+--- was unsuccessful returns `nil`.
+--- @field request_sync fun(method: string, params: table?, timeout_ms: integer?, bufnr: integer): {err: lsp.ResponseError|nil, result:any}|nil, string|nil err # a dictionary, where
+---
+--- Sends a notification to an LSP server.
+--- Returns: a boolean to indicate if the notification was successful. If
+--- it is false, then it will always be false (the client has shutdown).
+--- @field notify fun(method: string, params: table?): boolean
+---
+--- Cancels a request with a given request id.
+--- Returns: same as `notify()`.
+--- @field cancel_request fun(id: integer): boolean
+---
+--- Stops a client, optionally with force.
+--- By default, it will just ask the server to shutdown without force.
+--- If you request to stop a client which has previously been requested to
+--- shutdown, it will automatically escalate and force shutdown.
+--- @field stop fun(force?: boolean)
+---
+--- Runs the on_attach function from the client's config if it was defined.
+--- Useful for buffer-local setup.
+--- @field on_attach fun(bufnr: integer)
+---
+--- Checks if a client supports a given method.
+--- Always returns true for unknown off-spec methods.
+--- [opts] is a optional `{bufnr?: integer}` table.
+--- Some language server capabilities can be file specific.
+--- @field supports_method fun(method: string, opts?: {bufnr: integer?}): boolean
+---
+--- Checks whether a client is stopped.
+--- Returns: true if the client is fully stopped.
+--- @field is_stopped fun(): boolean
+local Client = {}
+Client.__index = Client
+
+--- @param cls table
+--- @param meth any
+--- @return function
+local function method_wrapper(cls, meth)
+  return function(...)
+    return meth(cls, ...)
+  end
+end
+
+--- @package
+--- @param id integer
+--- @param rpc vim.lsp.rpc.PublicClient
+--- @param handlers table<string,lsp.Handler>
+--- @param offset_encoding string
+--- @param config lsp.ClientConfig
+--- @return lsp.Client
+function Client.new(id, rpc, handlers, offset_encoding, config)
+  local name = config.name
+
+  --- @class lsp.Client
+  local self = {
+    id = id,
+    config = config,
+    handlers = handlers,
+    rpc = rpc,
+    offset_encoding = offset_encoding,
+    name = name,
+    _log_prefix = string.format('LSP[%s]', name),
```

Comment:
This looks like a great refactoring. I think the failing tests might be relevant to this change:
```
Error executing lua: ...er/work/neovim/neovim/runtime/lua/vim/lsp/client.lua:140: bad argument #2 to 'format' (string expected, got nil)
```

The code removed from `lsp.lua` and added here (moved) are slightly different, so it's not very easy to have a clear diff through web UI.

Reply:
Ah. I assumed because the other functionaltests passed that the failure was unrelated.

Looks like something unusual is happening with puc-lua. Can fix in a follow up.

Reply:
fixed in 5785c32